controllers/TOBASCOController.py

```diff
@@ -0,0 +1,250 @@
+#!/usr/bin/env python
+# -*- Mode: Python -*-
+# -*- encoding: utf-8 -*-
+# Copyright (c) Patrick Bartels <pckbls@gmail.com>
+
+# This file may be distributed and/or modified under the terms of
+# the GNU General Public License version 2 as published by
+# the Free Software Foundation.
+# This file is distributed without any warranty; without even the implied
+# warranty of merchantability or fitness for a particular purpose.
+# See "LICENSE" in the source distribution for more information.
+
+import math
+from utils_py.util import debug, format_bytes
+from BaseController import BaseController
+
+DEBUG = 1
+
+# This controller is an implementation of the TOBASCO Controller
+# (Threshold-Based Adaptation Scheme for on-Demand Streaming)
+# as described in "Adaptation Algorithm for Adaptive Streaming over HTTP"
+# by K. Miller, E. Quacchio, G. Gennari and A. Wolisz
+
+class TOBASCOController(BaseController):
+
+    def __init__(self):
+        super(TOBASCOController, self).__init__()
+
+        # shows whether we are in fast start mode
+        self.runningFastStart = True
+
+        # algorithm configuration parameters
+        self.conf = {
+            "B_min": 5,
+            "B_low": 10,
+            "B_high": 50,
+            "delta_beta": 1,
+            "delta_t": 5,
+            "alpha_1": 0.75,
+            "alpha_2": 0.33,
+            "alpha_3": 0.5,
+            "alpha_4": 0.75,
+            "alpha_5": 1.5 # choose 0.5 if a more conservative adaptation behavior is desired
+        }
+
+        # refer to isBuffering function for description
+        self.isBufferingHack = True
+
+        # the algorithm relies on past data, thus as an simple approach we simply save
+        # all feedback data in a history list
+        self.feedback_hist = []
+
+    def __repr__(self):
+        return '<TOBASCOController-%d>' % id(self)
+
+    def setPlayerFeedback(self, dict_params):
+        super(TOBASCOController, self).setPlayerFeedback(dict_params)
+
+        # append the data to our feedback history
+        self.feedback_hist.append(self.feedback)
+
+    def isBuffering(self):
```

Comment:
That's the main reason to leave the implementation of isBuffering to the controller :), I wouldn't call it a hack, it's just the way you define your "buffering" phase.